ecosystem/indexer-grpc/indexer-grpc-manager/src/metadata_manager.rs

```diff
@@ -0,0 +1,383 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use aptos_indexer_grpc_utils::timestamp_now_proto;
+use aptos_protos::{
+    indexer::v1::{
+        data_service_client::DataServiceClient, grpc_manager_client::GrpcManagerClient,
+        service_info::Info, FullnodeInfo, GrpcManagerInfo, HeartbeatRequest,
+        HistoricalDataServiceInfo, LiveDataServiceInfo, PingDataServiceRequest, ServiceInfo,
+    },
+    internal::fullnode::v1::{fullnode_data_client::FullnodeDataClient, PingFullnodeRequest},
+    util::timestamp::Timestamp,
+};
+use dashmap::DashMap;
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::atomic::{AtomicU64, Ordering},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use tonic::transport::channel::Channel;
+use tracing::trace;
+
+const MAX_NUM_OF_STATES_TO_KEEP: usize = 100;
```

Comment:
Adding some comments here would be helpful.

Reply:
Done